src/hardware/GPIOPin.h

```diff
@@ -0,0 +1,33 @@
+/**
+ * @file GPIOPin.h
+ * 
+ * @brief Abstraction of a GPIO pin 
+ */
+
+#pragma once
+
+#include <Arduino.h>
+
+enum PinType {
+    GPIO_OUTPUT,
+    GPIO_INPUT_STANDARD,
+    GPIO_INPUT_PULLUP,      // Internal pull-up resistor
+    GPIO_INPUT_PULLDOWN,    // Internal pull-down resistor
+    GPIO_INPUT_HARDWARE,    // External pull-up/pull-down resistor
+    GPIO_INPUT_ANALOG
+};
+
+class GPIOPin {
+    public:
+        GPIOPin(int pinNumber, PinType pinType);
+
+        void write(bool value);
+        void setPinType(PinType pinType);
+        int read();
```

Comment:
You could mark methods not changing the object as `const`:

```suggestion
        int read() const;
```

(same for many others) This makes it a bit more safe in cases where you *do not want to* alter anything. Then you can pass `const GPIOPin` and would get a compiler error if the code would try to ca.. e.g. `setPinType()`